server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java

```diff
@@ -168,6 +168,9 @@ protected boolean processField(RootObjectMapper.Builder builder, String fieldNam
                       }
                   ]
                 */
+                if (!(fieldNode instanceof List)) {
+                    throw new MapperParsingException("Dynamic template syntax error");
```

Comment:
I think it would be good to provide some more information the user what went wrong. In this case it would be great to tell the user know that `dynamic_templates` expects [an array of named objects](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-templates.html) or something along those lines.